### ISSUE TITLE
gcrane cp vs gcloud container images add-tag

### ISSUE THREAD

**Issue by erict-square** [google/go-containerregistry#905]
Hello this is an inquiry about `gcrane`. We've been hitting a problem where we need to copy images from a GCR in one project to a GCR in another project.

With `gcloud` CLI (With this [instruction](https://cloud.google.com/artifact-registry/docs/docker/copy-from-gcr#copy-gcloud)), we naively assumed that Google would do something smart and perform the underlying GCS movement strictly on Google's end, instead of copying and uploading the same image via the client.

We're wondering if `gcrane cp` does anything different. Does it behave similarly to `gcloud` where images need to be downloaded on the client before the upload?

**Comment by jonjohnsonjr** [comment]
tl;dr your assumption is correct as long as the source and destination have the same storage class and location.

The behavior of `gcloud container images add-tag` and `gcrane cp` are roughly equivalent, but it may not be entirely clear what's going on. I'll try to explain the flow, which should hopefully expose where my (or your) assumptions failed.

> we naively assumed that Google would do something smart and perform the underlying GCS movement strictly on Google's end, instead of copying and uploading the same image via the client.

This should be happening, but there are some limitations (that we may want to document better):

1. All interactions with GCR (via gcloud or gcrane) go through the [registry API](https://github.com/opencontainers/distribution-spec/blob/master/spec.md), so there is no magic happening anywhere. The only place we deviate is in the tags list response (see `gcrane ls --json`).
2. To trigger smart blob movement, you need to perform a [cross-repository mount](https://github.com/opencontainers/distribution-spec/blob/master/spec.md#mounting-a-blob-from-another-repository) of the blob. If that fails, we'll fall back to the slow path of downloading and uploading the blob.
3. In order for the mount to work:
    1. The identity performing the push needs to have read access to the source repository and write access to the destination repository.
    2. The repositories need to be in the same region, with the same storage class.
    3. (Really, all of the caveats for the [cloud storage rewrite](https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite) API apply.)

Notably, if you're trying to `gcrane cp us.gcr.io/$SRC_PROJECT/$IMAGE gcr.io/$DST_PROJECT/$IMAGE`, _this will not work_. Really, any hostname difference here will be a problem.

If you do need to copy images between regions, I'd recommend doing it within a GCE VM (or something) to keep the traffic internal, which should be much faster than e.g. roundtripping through residential internet.

Given all that, if the blobs still aren't mounting successfully, I'd be happy to help debug further. I find `gcrane cp -v ...` a good place to start for figuring out what's going wrong. If you see a blob POST return a 202, that's a failed mount.
